networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/peers/PeerChainValidator.java

```diff
@@ -0,0 +1,197 @@
+        .whenComplete(
+            (isValid, error) -> {
+              if (error != null) {
+                LOG.debug("Unable to validate peer's chain, disconnecting: " + peer, error);
+                peer.sendGoodbye(GoodbyeMessage.REASON_UNABLE_TO_VERIFY_NETWORK);
+                return;
+              } else if (!isValid) {
+                // We are not on the same chain
+                LOG.trace("Disconnecting peer on different chain: {}", peer);
+                peer.sendGoodbye(GoodbyeMessage.REASON_IRRELEVANT_NETWORK);
+                return;
+              }
+              LOG.trace("Validated peer's chain: {}", peer);
+              peer.markChainValidated();
+            });
+  }
+
+  private CompletableFuture<Boolean> checkRemoteChain() {
+    checkState(peer.hasStatus(), "Peer must have an associated status.");
+    final StatusData peerStatus = peer.getStatus();
+
+    final CompletableFuture<Boolean> isChainValid = new CompletableFuture<>();
+
+    // Check fork compatibility
+    Bytes4 expectedFork = storageClient.getForkAtSlot(peerStatus.getHeadSlot());
+    if (!Objects.equals(expectedFork, peerStatus.getHeadForkVersion())) {
+      LOG.trace(
+          "Peer's fork ({}) differs from our fork ({}): {}",
+          peerStatus.getHeadForkVersion(),
+          expectedFork,
+          peer);
+      isChainValid.complete(false);
+      return isChainValid;
+    }
+
+    // If we haven't reached genesis, accept our peer at this point
+    if (storageClient.isPreGenesis()) {
+      LOG.trace("Validating peer pre-genesis, skip finalized block checks for peer {}", peer);
+      isChainValid.complete(true);
+      return isChainValid;
+    }
+
+    // Check whether finalized checkpoints are compatible
+    final Checkpoint finalizedCheckpoint = storageClient.getStore().getFinalizedCheckpoint();
+    final UnsignedLong finalizedEpoch = finalizedCheckpoint.getEpoch();
+    final UnsignedLong remoteFinalizedEpoch = peerStatus.getFinalizedEpoch();
+
+    if (finalizedEpoch.compareTo(remoteFinalizedEpoch) == 0) {
+      final boolean chainsAreConsistent =
+          verifyFinalizedCheckpointsAreTheSame(finalizedCheckpoint, peerStatus);
+      isChainValid.complete(chainsAreConsistent);
+    } else if (finalizedEpoch.compareTo(remoteFinalizedEpoch) > 0) {
+      // We're ahead of our peer, check that we agree with our peer's finalized epoch
+      verifyPeersFinalizedCheckpointIsCanonical(peerStatus, isChainValid);
+    } else {
+      // Our peer is ahead of us, check that they agree on our finalized epoch
+      verifyPeerAgreesWithOurFinalizedCheckpoint(finalizedCheckpoint, peerStatus, isChainValid);
+    }
+    return isChainValid;
+  }
+
+  private boolean verifyFinalizedCheckpointsAreTheSame(
+      Checkpoint finalizedCheckpoint, StatusData peerStatus) {
+    return Objects.equals(finalizedCheckpoint.getRoot(), peerStatus.getFinalizedRoot());
+  }
+
+  private void verifyPeersFinalizedCheckpointIsCanonical(
```

Comment:
(nit) I would name this method and  `verifyPeerAgreesWithOurFinalizedCheckpoint` to something like `verifyChainsMatchWhenPeerChainIsLonger` and `verifyChainsMatchWhenOurChainIsLonger`